[SPARK-13547] [SQL] [WEBUI] Add SQL query in web UI's SQL Tab

nblintao opened the pull request:
## What changes were proposed in this pull request?

For the web UI, this PR adds the raw SQL query text for each query in the new column "SQL Text" in the SQL Tab. To achieve this, it also passes the SQL text from the parser to the execution UI data, through the logical plan.

If the query has no related query text, show a blank in that field. If there is no query in the table has a query text, not show the column.

If the query text is long (i.e. longer than 140 chars), show the short abstract and display the full version on demand by the link "+more".

If the query text is too long (i.e. longer than 1000 chars), it will be truncated after parsing. This is to avoid downing browsers or web UI servers by running out of memory.

Also, a tooltip is added on the header "SQL Text", saying `Shows 140 characters by default. Click "+more" to see more. Long texts are truncated to 1000 characters. Left blank when the query was not issued by SQL.`.
## How was this patch tested?

Tested by trying multiple SQL queries.

A screenshot of the SQL Tab:
![sqltextwithtooltip](https://cloud.githubusercontent.com/assets/5558370/16863133/5ddb500c-4a15-11e6-9b08-e4ac160a0cd5.png)
(Query 3 is identical to Query 2, but its SQL text is shown in detail by clicking "+more")

nblintao committed:
add SQL query texts in the web UI

SparkQA commented:
**[Test build #62180 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62180/consoleFull)** for PR 14158 at commit [`626f3f7`](https://github.com/apache/spark/commit/626f3f74fb3fa09dd781c4e11d7412776fe06970).

SparkQA commented:
**[Test build #62180 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62180/consoleFull)** for PR 14158 at commit [`626f3f7`](https://github.com/apache/spark/commit/626f3f74fb3fa09dd781c4e11d7412776fe06970).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62180/
Test FAILed.

ajbozarth commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala:
```diff
@@ -103,6 +104,9 @@ private[ui] abstract class ExecutionTable(
         {descriptionCell(executionUIData)}
       </td>
+      <td>
+        {executionUIData.sqlText.getOrElse("-")}
```
Is there a reason to insert a "-" rather than leave this blank in the case it's not available?

nblintao replied:
I've just checked and found out that the common practice in the Spark Web UI is to use a blank instead of a hyphen. I'll fix. Thanks!

ajbozarth commented:
Checked it out and gave it a test run and it looks good, just one issue. When the user only uses the dataframes api it will never show any SQL Text (only "-"), should we check if there is no SQL Text to display and not show the column in that case?

nblintao commented:
Thanks, @ajbozarth. That's a great point. When no execution in a table has SQL text, this column shouldn't be shown. I'll fix this.

nblintao committed:
optimize for long query strings & not show the column when no query string

SparkQA commented:
**[Test build #62207 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62207/consoleFull)** for PR 14158 at commit [`beba227`](https://github.com/apache/spark/commit/beba2279aa9505a696189755a190f76350277d35).

nblintao commented:
**In the commit above, I made three adjustments:**

1. If the query has no related query text, show a blank in that field.

1. If there is no query in the table has a query text, not show the column.

1. If the query text is too long (i.e. longer than 140 chars), show the short abstract and display the full version on demand.

**A snapshot of the SQL Tab:**
![longsqltext](https://cloud.githubusercontent.com/assets/5558370/16790563/9c4e638c-487c-11e6-9478-e3e4c804f182.png)
(The detailed SQL text for Query 4 shown after clicking "+more")

**The following need to be discussed and implemented (suggested by @yhuai):**

1. Add a ToolTip on "SQL Text", saying "Blank if the query is not issued by SQL" (Please help check this sentence if you are a native English speaker).

2. For now, we could show the abstract of the string at the front end. But we still need to truncate the query texts which are toooooo long (maybe longer than 1000 bytes?). This is to avoid downing browsers or web UI servers by running out of memory.

SparkQA commented:
**[Test build #62207 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62207/consoleFull)** for PR 14158 at commit [`beba227`](https://github.com/apache/spark/commit/beba2279aa9505a696189755a190f76350277d35).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62207/
Test PASSed.

felixcheung commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala:
```diff
@@ -146,11 +158,35 @@ private[ui] abstract class ExecutionTable(
 
+  private def sqlTextCell(sqlText: String): Seq[Node] = {
+    // Only show a limited number of characters of sqlText by default when it is too long
+    val maxLength = 140
```
how about a button to copy full query text into clipboard?

ajbozarth replied:
Is this really necessary? The text can be easily highlighted and copied.

felixcheung replied:
.. for very long query that wouldn't fit as per this [comment](https://github.com/apache/spark/pull/14158#issuecomment-232252878)

ajbozarth replied:
Theres a "+more" button that expands to show the full query

ajbozarth replied:
nvm, I understand you now, if we end up truncating the text at 1000 chars as proposed then yes this may be a useful feature, but if the text is available to copy wouldn't the size issue still exist?

felixcheung replied:
you mean that it could be too long for clipboard? I don't think that would be an issue...

nblintao replied:
Thanks, @felixcheung. When I am looking at the discussions again today, I find I am more clear about your suggestions. But actually, truncating very long queries is not because they are not visually pleasant to be displayed on the page, but it may take too many resources to store and transfer such a long unlimited string. So the queries longer than 1000 chars are truncated after parsing, which will never be recovered. So the idea of clipboard may not be used to resolve it.

But still, I believe the clipboard is a good idea. I do think we could extend the limit to 10k chars instead of 1k if we add a clipboard later.

ajbozarth commented:
Checked out your changes and I like the updates, plus I think the column looks better on the end. Addressing the questions you raised: i agree we should truncate long queries (1000 chars sounds like a decent number), we should make it clear they were truncated though (maybe with a "..."?) I also think a tool tip would be good. Maybe something like: "Left blank when the query was not issued by SQL"?

nblintao commented:
@ajbozarth @felixcheung Thanks for viewing and discussing! I plan to implement the two updates I mentioned above. The idea of clipboard sounds awesome, but I think we might need more discussion about this new function.

nblintao committed:
truncate long text & add tool tip

SparkQA commented:
**[Test build #62364 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62364/consoleFull)** for PR 14158 at commit [`41c2daa`](https://github.com/apache/spark/commit/41c2daa19a4b4dc340f6345e2624fd269565638b).

nblintao commented:
Updated by truncating long texts and adding a tooltip.
The detail description and the screenshot at https://github.com/apache/spark/pull/14158#issue-165127460 is also updated.

SparkQA commented:
**[Test build #62364 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62364/consoleFull)** for PR 14158 at commit [`41c2daa`](https://github.com/apache/spark/commit/41c2daa19a4b4dc340f6345e2624fd269565638b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62364/
Test PASSed.

ajbozarth commented:
I'll try to take a look at this tomorrow

ajbozarth commented:
The tooltip looks odd centered on the column instead of the title text but I believe that's how the whole Web UI works, other than that nit LGTM

ajbozarth commented:
@srowen @JoshRosen what do you think of @nblintao 's pr? Some of my team really likes the addition and would love to see it merged

ajbozarth commented:
@nblintao Could you rebase? I'd like to try and see if we can get this reviewed and merged.

markhamstra commented:
Please be certain that this works well even for very large queries.  They are not commonplace, but I know that Spark SQL does sometimes get asked to handle SQL queries that are hundreds of lines  long.

ajbozarth commented:
Thanks @markhamstra that was already addressed earlier in review

markhamstra commented:
It's great if it is already addressed, but I just didn't see anything explicit in the discussion or examples that showed any query of the magnitude that I am talking about.  Machine-generated queries can be staggeringly large (e.g. I've seen queries that were so long and complicated that they took Spark SQL more than 9 minutes just to parse), and this UI enhancement must be prepared to handle those.

nblintao commented:
Thanks, @ajbozarth. Sure, I'd love to rebase it and get it merged. But I am really busy with my mid-terms and interviews recently. I'll take a look tonight and fix it if it's easy to do.

I really appreciate your concern about my pull requests.

nblintao commented:
@markhamstra Thanks. Here is my related explanation at the top of this PR.

> If the query text is too long (i.e. longer than 1000 chars), it will be truncated. This is to avoid downing browsers or web UI servers by running out of memory.

Is this what you mean?

markhamstra commented:
@nblintao I really haven't reviewed these changes closely enough to have a specific complaint or concern in mind, but I'm more concerned about what happens when you ask to see "more" when that "more" could be several tens of thousands of characters, not just the few thousand that many would consider to be a very long SQL query.

nblintao commented:
@markhamstra Queries longer than 1000 chars are truncated to 1000 chars after the parsing. "+more" will only expand 140 chars to at most 1000 chars.

markhamstra commented:
@nblintao Got it; thanks.  There may be distinct queries that will be entirely the same within the first 1000 characters, but that's just the nature of working with these very large queries -- there are lots of things that make them difficult, but it sounds like you've made sure not to make things difficult for more normal queries as a consequence, so that's all good.

nblintao commented:
@ajbozarth I finally have a chance to rebase it in the winter break. Could you please have a look? Thanks!

SparkQA commented:
**[Test build #70700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70700/testReport)** for PR 14158 at commit [`83cbb58`](https://github.com/apache/spark/commit/83cbb588320d945d34dfd28e49eba90bcf63b064).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70700/
Test FAILed.

nblintao commented:
retest this please

SparkQA commented:
**[Test build #70704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70704/testReport)** for PR 14158 at commit [`83cbb58`](https://github.com/apache/spark/commit/83cbb588320d945d34dfd28e49eba90bcf63b064).

SparkQA commented:
**[Test build #70704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70704/testReport)** for PR 14158 at commit [`83cbb58`](https://github.com/apache/spark/commit/83cbb588320d945d34dfd28e49eba90bcf63b064).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

ajbozarth commented on core/src/main/scala/org/apache/spark/ui/UIUtils.scala:
```diff
@@ -326,7 +336,16 @@ private[spark] object UIUtils extends Logging {
+          <th width={colWidthAttr} class={getClass(x._2)}>{getHeaderContent(x._1)}</th>
+        } else {
+          val toolTip = toolTipOption.get
+          // scalastyle:off line.size.limit
```
you shouldn't need this, scala can handle line wrapping for `Node` return values

nblintao replied:
Fixed. Thank you!

ajbozarth commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:
```diff
@@ -50,13 +50,29 @@ abstract class AbstractSqlParser extends ParserInterface with Logging {
+    // Truncate the text to avoid downing browsers or web UI servers by running out of memory.
+    val maxLength = 1000
+    val suffix = " ... (truncated)"
+    val truncateLength = maxLength - suffix.length
```
Is this really necessary? Given the reasoning behind the max length I don't see a reason to include extra lines of code to include the suffix as part of the max length.

nblintao replied:
I think either way is okay. Here, I am considering keeping the text displayed (including suffix) less than 1000 chars.

ajbozarth reviewed (commented):
A couple of comments, but from a UI side this looks ok to me. @gatorsmile What do you think of this from the SQL side? Also @srowen and @vanzin second opinion on the UI side?

HyukjinKwon commented:
Hi @nblintao, is it still active?

nblintao commented:
@HyukjinKwon Thanks for reminding. I am finals this week. I will fix this after 15th.

SparkQA commented:
**[Test build #77510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77510/testReport)** for PR 14158 at commit [`69180bd`](https://github.com/apache/spark/commit/69180bd5b5b21725ff1e498e98690bc261f079f7).

nblintao commented:
I have just rebased. @ajbozarth @HyukjinKwon  @gatorsmile @srowen @vanzin

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -258,6 +258,9 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
   def refresh(): Unit = children.foreach(_.refresh())
+
+  // Record the original sql text in the top logical plan for checking in the web UI.
+  var sqlText: Option[String] = None
```
Using `var` for this should be avoided.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:
```diff
@@ -65,13 +65,29 @@ abstract class AbstractSqlParser extends ParserInterface with Logging {
+      }
     }
+    logicalPlan.sqlText = Some(truncatedSqlText)
+    logicalPlan
```
The solution in this PR looks intrusive to me. If we really want to store the original sql text, we can add it into the QueryExecution. The value can be initialized when we build the QueryExecution

nblintao replied:
Thanks for pointing it out. I agree that `QueryExecution` is a better place for original SQL text. I have updated my code accordingly. Could you please have a look?

SparkQA commented:
**[Test build #77510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77510/testReport)** for PR 14158 at commit [`69180bd`](https://github.com/apache/spark/commit/69180bd5b5b21725ff1e498e98690bc261f079f7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77510/
Test PASSed.

nblintao committed:
Remove unnecessary scala style

SparkQA commented:
**[Test build #77518 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77518/testReport)** for PR 14158 at commit [`114401a`](https://github.com/apache/spark/commit/114401a630650623c7c311bf753d4422d98e1550).

SparkQA commented:
**[Test build #77518 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77518/testReport)** for PR 14158 at commit [`114401a`](https://github.com/apache/spark/commit/114401a630650623c7c311bf753d4422d98e1550).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77518/
Test FAILed.

nblintao committed:
Add back Scala style

nblintao committed:
Store sql text in QueryExecution instead of logical plan

SparkQA commented:
**[Test build #77523 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77523/testReport)** for PR 14158 at commit [`ac96aaa`](https://github.com/apache/spark/commit/ac96aaa995330a3b6849cbbfa68222a2e90d0e15).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77523/
Test FAILed.

jiangxb1987 commented:
retest this please

SparkQA commented:
**[Test build #78291 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78291/testReport)** for PR 14158 at commit [`ac96aaa`](https://github.com/apache/spark/commit/ac96aaa995330a3b6849cbbfa68222a2e90d0e15).

AmplabJenkins commented:
Merged build finished. Test FAILed.

shaneknapp commented:
i'll retrigger this once jenkins is back up.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #78299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78299/testReport)** for PR 14158 at commit [`ac96aaa`](https://github.com/apache/spark/commit/ac96aaa995330a3b6849cbbfa68222a2e90d0e15).

SparkQA commented:
**[Test build #78299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78299/testReport)** for PR 14158 at commit [`ac96aaa`](https://github.com/apache/spark/commit/ac96aaa995330a3b6849cbbfa68222a2e90d0e15).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class QueryExecution(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78299/
Test FAILed.

jiangxb1987 commented:
retest this please

SparkQA commented:
**[Test build #78332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78332/testReport)** for PR 14158 at commit [`ac96aaa`](https://github.com/apache/spark/commit/ac96aaa995330a3b6849cbbfa68222a2e90d0e15).

SparkQA commented:
**[Test build #78332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78332/testReport)** for PR 14158 at commit [`ac96aaa`](https://github.com/apache/spark/commit/ac96aaa995330a3b6849cbbfa68222a2e90d0e15).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class QueryExecution(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78332/
Test FAILed.

HyukjinKwon commented:
gentle ping @nblintao, is this PR active? If so, I guess the test failure should be fixed if related.

nblintao commented:
@HyukjinKwon It's still active. I'll fix it when I'm available. Thanks.

HyukjinKwon commented:
hey @nblintao, do you maybe happened to have some time to continue this one?

nblintao commented:
@HyukjinKwon Sorry for the delay. I'm busy looking for jobs these days. I'll try my best to fix it in October. Thank you for reminding me!

jiangxb1987 commented:
I'm going to close this PR because it goes stale, please feel free to reopen it or open another PR if anyone have more thoughts on this issue.
